tests: port a bunch of kernel test cases to userspace

tyrantlizard77 opened the pull request:
Objects tested:
mutex
semaphore
alert
msgq
stack
thread

Timer tests still WIP. Pipe API blocked by #4366

Commit:
tests: mutex: run in user mode

Main thread grants itself access to objects it or its children need
and does the rest of the test case in user mode.

Statically defined threads now all run in user mode, with permissions
granted via K_THREAD_ACCESS_GRANT().

Added userspace tag to testcase.yaml.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: alert_api: run in user mode

main.c and test_alert_contexts.c merged.

User threads can't look inside the alert structures, so an extra
variable 'htype' introduced to track expectations for any given
alert object in alert_recv().

Alert objects have to be initialized by supervisor threads since
they register callbacks. An array of toplevel alert objects created
and initialized in test_main(), replacing the ones that used to
live on thread stacks.

Added userspace tag to testcase.yaml

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: semaphore: run some threads in user mode

main.c and test_sema_contexts.c merged

userspace tag added to testcase.yaml

stack-allocated semaphore in test_sema_thread2thread now just uses
the global semaphore with the same name.

ISR tests run in supervisor mode.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: msgq: run in user mode when possible

Unnecessary k_thread_abort() removed from test_msgq_purge_when_put.

A single global msgq object is now shared instead of being declared
on thread stacks, except for an ISR test case which has had its
semaphore renamed.

Moved k_sem_init() call from msgq_thread() to test_msgq_thread()
to fix a race condition.

userspace tag added to testcase.yaml.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: stack_api: run in user mode when possible

All non ISR tests now run in user mode.

userspace added to testcase.yaml tag list.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: custom_data: run in user mode

main.c and test_customdata_api.c merged.

Preemptive priority case now run in user mode.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: lifecycle_api: run in user mode

Some unnecessary k_thread_abort() removed.

userspace tag added to testcase.yaml.

Suspend/resume, spawn_forever, and spawn_priority tests remain in
supervisor mode due to the priority requests they make.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: thread_init: run in user mode

main.c and test_thread_init.c merged.

All tests which don't require cooperative priorities now running in
user mode.

Userspace tag added to testcase.yaml.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: obj_validation: add to userspace tests

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Note to reviewers: this patch is a good place to request changes to the application-facing APIs for userspace before they get ground into stone.

andyross reviewed (approved):
Seems sane.  One whine is that the pattern used here of calling k_thread_user_mode_enter() means that the traditional/non-userspace code now runs with two extra frames on the stack, needlessly.

Might be nice to have a framework to have a "userspace thread entry" that interoperated in an obvious way with main() so that we didn't need to make the extra calls.  Maybe just as part of the test framework.

ramakrishnapallala commented on tests/kernel/threads/custom_data/custom_data_api/src/test_customdata_api.c:
```diff
@@ -58,9 +58,20 @@ void test_customdata_get_set_preempt(void)
+	k_thread_access_grant(k_current_get(), &tdata, tstack, NULL);
+
+	ztest_test_suite(test_customdata_api,
+		ztest_unit_test(test_customdata_get_set_coop),
```
any reason for not running the COOP test in user mode?

andrewboie replied:
user threads can't create new threads that are of greater priority than they are.

ramakrishnapallala commented on tests/kernel/threads/lifecycle/lifecycle_api/src/main.c:
```diff
@@ -25,21 +25,27 @@ extern void test_threads_cancel_started(void);
-			 ztest_unit_test(test_threads_spawn_delay),
+			 ztest_user_unit_test(test_threads_spawn_delay),
 			 ztest_unit_test(test_threads_spawn_forever),
 			 ztest_unit_test(test_threads_suspend_resume_cooperative),
```
just out of curiosity, what will be the impact if we run the spawn and suspend tests in user mode? i see that suspend/resume just passing the priority number it wants, how does this impact the user mode execution.?

andrewboie replied:
they will fail, k_thread_priority_set() only lets threads downgrade priority, not upgrade it

ramakrishnapallala commented on tests/kernel/threads/lifecycle/thread_init/src/test_thread_init.c:
```diff
@@ -194,3 +197,17 @@ void test_kinit_coop_thread(void)
+			 ztest_user_unit_test(test_kdefine_preempt_thread),
+			 ztest_user_unit_test(test_kdefine_coop_thread),
+			 ztest_user_unit_test(test_kinit_preempt_thread),
+			 ztest_unit_test(test_kinit_coop_thread));
```
same doubt again, why coop threads can't run in user mode?
also why coop thread test at 209 is run in user mode?

andrewboie replied:
this test was skipped, user threads can't create new threads that are of greater priority than they are

the other coop test works fine

nashif reviewed (changes requested):
we need to test the same with userspace not being enabled, I know we will get this on boards of other architectures that do not have userspace enable by default (it is for qemu), but a test strategy for the APIs with and without userspace on the same target needs to be addressed.

Commit:
x86: add qemu_x86_nommu board target

This will run all tests with the memory management unit disabled.
This means no hardware-based stack protection or user threads.

qemu_x86_iamcu now runs with all MMU features enabled.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: set default CI targets

The IAMCU variant does not need to be the default, this calling
convention was to support the discontinued Quark platforms.

qemu_x86 and qemu_x86_nommu now are set as default.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
per @nashif feedback, added more patches to add "qemu_x86_nommu" build target to exercise x86 test cases with no memory management features

nashif commented on boards/x86/qemu_x86/qemu_x86_nommu.yaml:
```diff
@@ -0,0 +1,9 @@
+  - zephyr
+supported:
+  - pci
+  - netif:slip
```
if you enable this as default then we will have full coverage of userspace/non-userspace for the tests, it will increase the normal sanitycheck run from 1077 to 1237 but this acceptable for making sure we do not break any APIs for non-userspace usage

andrewboie commented:
updated, qemu_x86 and qemu_x86_nommu now set with "default"
